database.rst

```diff
@@ -0,0 +1,181 @@
+
+    systemctl start postgresql
+    systemctl enable postgresql
+
+Sanity-check that the daemon is listening on all interfaces::
+
+    netstat -an | grep 5432
+
+This should should show postgres listening on ``0.0.0.0:5432``.
+
+Connecting
+----------
+
+On your local workstation, connect to your Posgres VM's IP address. Here's an
+example connecting to Postgres on RHEL 7::
+
+  psql -h 192.168.122.124 -U postgres
+
+  postgres=# select VERSION();
+   PostgreSQL 9.2.23 on x86_64-redhat-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16), 64-bit
+   (1 row)
+
+Compare this to production composedb environment (currently RHEL 5)::
+
+  psql -h (hostname) -U (username) compose -W -c "select VERSION();"
+   PostgreSQL 8.2.14 on x86_64-redhat-linux-gnu, compiled by GCC gcc (GCC) 4.1.2 20080704 (Red Hat 4.1.2-46)
+
+
+The ``\q`` command will quit the ``psql`` prompt.
+
+Create the application user
+---------------------------
+
+On your local workstation, you can use the ``createuser`` shell utility
+(``postgresql`` package) to create the "compose_ro" user account::
+
+  createuser -h 192.168.122.124 -U postgres -P --no-createdb --no-createrole compose_ro
+
+``createuser`` will prompt you to set a password for this new user.
+
+Create the application database
+-------------------------------
+
+On your local workstation, you can use the ``createdb`` shell utility
+(``postgresql`` package) to create the "compose" database::
+
+  createdb -h 192.168.122.124 -U postgres --encoding SQL_ASCII compose
+
+On RHEL 7, you must append ``--template template0`` to this command.
+
+Note, the encoding is important, because this is what is set on the prod
+server. *If* you forgot to set ``--encoding`` during ``createdb`` above, you
+can set it later::
+
+  update pg_database set encoding = pg_char_to_encoding('SQL_ASCII') where datname = 'compose';
+
+If you don't set the encoding here, you cannot import data dumps from
+the production server.
+
+Dumping and restoring tables from production
+--------------------------------------------
+
+products.py needs the following tables::
+
+  match_versions
+  overrides
+  packages
+  products
+  tree_packages
+  tree_product_map
+  trees
+
+You can use ``db_dump`` to obtain these from production. Here are the file
+sizes from ``db_dump``'s custom compressed format::
+
+    697  match_versions
+    666K overrides.backup
+    157M packages.backup
+    31K  products.backup
+    1.3G tree_packages.backup
+    7.1M trees.backup
+
+Note: the ``tree_packages`` table is 1.3GB compressed (unknown size
+uncompressed). It took 18 minutes to dump from prod over my VPN connection
+(1.25MB/s).
+
+I've not yet been able to restore this ``tree_packages`` table to a development
+VM. On RHEL 5, ``pg_restore`` churns for an hour and then OOMs even with 4GB
+RAM. On RHEL 7 ``pg_restore`` does not leak memory, but I kept running out of
+disk space in ``/var/lib/pgsql/data``. It took up 50GB of space before I gave
+up.
```

Comment:
Woah, that's crazy.

Reply:
Yeah, I need to retry - I'm still shuffling stuff around on my laptop so I have enough space :cry:

Reply:
Just did today: `docker image prune && docker volume prune`

It saved me few GiBs.

Reply:
Yeah, I still cannot import this table with `pg_restore`, even with 6GB RAM and 4vCPUs, and tuning postgres to use less memory. Are you ok merging this as-is, and we'll come up with some better solution later?

Reply:
Sure, let's merge it.